File: leetcode/331-verify-preorder-serialization-of-a-binary-tree.py
class Solution(object):
    def isValidSerialization(self, preorder):
        """
        :type preorder: str
        :rtype: bool
        """
        null = False
        finish = False
        cnt = []
        for c in preorder + ',':
            if ',' == c:
                if not null: cnt += [0]
                null = False
            else:
                if finish: return False
                if '#' == c:
                    null = True
                    while len(cnt) > 0:
                        cnt[-1] += 1
                        if (2 == cnt[-1]): cnt.pop()
                        else: break
                    finish = len(cnt) == 0
        return len(cnt) == 0

File: leetcode/331-verify-preorder-serialization-of-a-binary-tree.cpp
class Solution
{
 public:
  bool isValidSerialization(string preorder)
  {
    vector<int> cnt;
    bool null = false, finish = false;
    for (char c : preorder + ',') {
      if (',' == c) {
        if (!null) cnt.push_back(0);
        null = false;
      } else {
        if (finish) return false;
        if ('#' == c) {
          null = true;
          while (!cnt.empty() && 2 == ++cnt.back())
            cnt.pop_back();
          finish = cnt.empty();
        }
      }
    }
    return cnt.empty();
  }
};
